[Filters] Add an internal feature key for enabling GraphicsContext filters

shallawa opened the pull request:
#### 74b151e788b03d24ea9ce1a2fc876a01d6589f14
<pre>
[Filters] Add an internal feature key for enabling GraphicsContext filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=248375">https://bugs.webkit.org/show_bug.cgi?id=248375</a>
rdar://102689264

Reviewed by Sam Weinig.

This key will be used for enabling GraphicsContext filters giving that all the
FilterEffects in the filter chain support CoreGraphics filters.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::style const):
(WebCore::GraphicsContext::setStyle):
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::GraphicsContextState::mergeLastChanges):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp.
(WebCore::FilterStyleTargetSwitcher::FilterStyleTargetSwitcher):
(WebCore::FilterStyleTargetSwitcher::willDrawSourceImage):
(WebCore::FilterStyleTargetSwitcher::didDrawSourceImage):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):

Canonical link: <a href="https://commits.webkit.org/257055@main">https://commits.webkit.org/257055@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a29730b7fa8563b047c536f69b2f36a081f0869

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97731 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6974 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30899 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/107222 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167490 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/7387 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35743 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/90117 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103873 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103374 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/5544 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84363 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32509 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87418 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/89173 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75429 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88629 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/966 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/20607 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/84306 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/956 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22102 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28525 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4848 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5778 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/44563 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/87138 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/2211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41497 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19549 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
```diff
@@ -0,0 +1,58 @@
+
+namespace WebCore {
+
+FilterStyleTargetSwitcher::FilterStyleTargetSwitcher(GraphicsContext&, Filter& filter, const FloatRect &sourceImageRect)
```
`const FloatRect &sourceImageRect` -> `const FloatRect& sourceImageRect` (& on wrong side)

weinig commented on Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
```diff
@@ -0,0 +1,58 @@
+FilterStyleTargetSwitcher::FilterStyleTargetSwitcher(GraphicsContext&, Filter& filter, const FloatRect &sourceImageRect)
+    : FilterTargetSwitcher(filter)
+{
+    m_filterStyles = filter.createFilterStyles(sourceImageRect);
```
This should be in the initialize list.

weinig commented on Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
```diff
@@ -0,0 +1,58 @@
+    m_filterStyles = filter.createFilterStyles(sourceImageRect);
+}
+
+void FilterStyleTargetSwitcher::willDrawSourceImage(GraphicsContext&  destinationContext, const FloatRect&)
```
`GraphicsContext&  destinationContext` - too many spaces.

weinig commented on Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp:
```diff
@@ -0,0 +1,58 @@
+
+void FilterStyleTargetSwitcher::didDrawSourceImage(GraphicsContext& destinationContext)
+{
+    for (int size = m_filterStyles.size(); size; --size) {
```
Since you are not using the m_filterStyles here, going backwards doesn't really matter, but if you want to keep that part incase anyone ever uses the style in the loop I would recommend:

```c++
for (auto& filterStyle : makeReverseRange(m_filterStyles) {
```

shallawa replied:
Done but I had to use 

```
for ([[maybe_unused]] auto& filterStyle : makeReversedRange(m_filterStyles)) {
```

because filterStyle is not used in the loop body.

shallawa committed:
[Filters] Add an internal feature key for enabling GraphicsContext filters
https://bugs.webkit.org/show_bug.cgi?id=248375
rdar://102689264

Reviewed by Sam Weinig.

This key will be used for enabling GraphicsContext filters giving that all the
FilterEffects in the filter chain support CoreGraphics filters.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::preferredFilterRenderingModes const):
* Source/WebCore/platform/graphics/GraphicsContext.h:
(WebCore::GraphicsContext::style const):
(WebCore::GraphicsContext::setStyle):
* Source/WebCore/platform/graphics/GraphicsContextState.cpp:
(WebCore::GraphicsContextState::mergeLastChanges):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.cpp: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp.
(WebCore::FilterStyleTargetSwitcher::FilterStyleTargetSwitcher):
(WebCore::FilterStyleTargetSwitcher::willDrawSourceImage):
(WebCore::FilterStyleTargetSwitcher::didDrawSourceImage):
* Source/WebCore/platform/graphics/filters/FilterStyleTargetSwitcher.h: Copied from Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp.
* Source/WebCore/platform/graphics/filters/FilterTargetSwitcher.cpp:
(WebCore::FilterTargetSwitcher::create):

Canonical link: https://commits.webkit.org/257055@main
